### ISSUE TITLE
Failed Linux Ubuntu Trusty installation.

### ISSUE THREAD

**Issue by gorfreed** [sandstorm-io/sandstorm#2473]
Like many others, I am receiving the "Your kernel does not appear to be compiled for..." kernel error when trying to install Sandstorm on my Ubuntu server. This seems to be a widespread issue (I am trying to install the software for months now, nobody of my admin colleagues has managed to succeed in it either) but forum post or support ticket I read about it seems to end nowhere! So I need to figure out by myself if Sandstorm simply cannot be installed on a root virtual server running Ubuntu 14.04 (which would be a shame). Please let me know what to need - I can tell you right away though that it is a pretty standard configuration I am working with. 

Thanks a lot for your time!

**Comment by gorfreed** [comment]
Yes, I have read all documentation on the support pages regarding this issue.

https://docs.sandstorm.io/en/latest/administering/faq/#enabling-user-namespaces

I have no clue what the link to the Debian bug report is supposed to tell me and I cannot find the folder of the command I am supposed to run.

# sysctl -w kernel.unprivileged_userns_clone=1
# echo 'kernel.unprivileged_userns_clone = 1' >> /etc/sysctl.conf"

**Comment by kentonv** [comment]
Hi @gorfreed,

We definitely want to help you get Sandstorm installed. I regularly test installing Sandstorm on Ubuntu 14.04 on Digital Ocean and have not observed any problems, so there must be something else going on here.

What hosting provider hosts the server you are trying to use? What is the output of `uname -a` on your machine? We've found that sometimes hosting providers inject their own kernel builds rather than using the one from the distro, so that could be part of the problem. Sometimes the provider has alternate kernel options that work.

Also, are you attempting to install inside Docker / another container engine, or are you installing on a normal non-containerized system?

**Comment by gorfreed** [comment]
Dear Kenton,

Thank you for the prompt reply!

Here is the missing info:

*uname -a result (with root):*
Linux euve96982 3.13.0-042stab113.11 #1 SMP Fri Dec 18 17:32:04 MSK 2015
x86_64 x86_64 x86_64 GNU/Linux

*Hosting Provider:*
I am trying to deploy Sandstorm on my private virtual server hosted at
server4you.net. You can check the specs of the server here
<http://www.server4you.net/virtual-server> (it's the one with the
preselected "best pick" configuration). I am trying to install on a normal
non-containerized system. I checked the server interface of the hosting
provider for configuration options related to OS/kernel and can confirm
that there aren't any. If we can get the software running with an alternate
kernel option, I would therefore have to request the feature change from
Server4You support. To increase my chances of success here, it would be
immensely helpful to know what to ask these guys for.

Please note that I experience the *very same issue* on the virtual server I
use for work, hosted by strato.de, Germany's biggest hosting provider

Please let me know if there are any other command outputs or logs that
would be helpful for your investigation.

I really appreciate your help, can't wait to get started with Sandstorm.io!
:)

Best,
Daniel

**Comment by kentonv** [comment]
Hi @gorfreed,

I looked at server4you and strato.de. Unfortunately, it looks to me like both of these companies sell "vservers", as in the Linux-VServer project. A "vserver" is actually not a virtual machine but rather a kind of container. Low-end hosting providers often use vservers to split a single virtual machine into many containers that can be used by different customers. This lets them offer very cheap servers -- but with poor performance and poor security.

Linux-VServer is actually an unofficial modification of the Linux kernel. Unfortunately, it is incompatible with the Linux kernel's official containerization features (which are, admittedly, newer than Linux-VServer).

Because of this, you may have found that you cannot run Docker on these servers -- it needs real Linux containerization, so is incompatible. Similarly, Sandstorm will not work, because it also needs the same containerization features Docker uses.

If you want to contact your hosting provider, probably the easiest thing to do is ask them how to get a server that can run Docker. If it can run Docker, then it can probably run Sandstorm as well. The people at your hosting provider probably get asked about Docker a lot, so will know how to answer.

We should probably improve the installer script so that it can detect vservers and explain this problem.

**Comment by gorfreed** [comment]
Thank you for looking into it!

That's very unfortunate, as it seems in order to use Sandstorm I would have
to switch to a more expensive server package. I am surprised about the
system requirements though. The current server is running Owncloud, Cozy
Cloud, Softaculous, Installatron, a pretty addon-heavy Wordpress, Redmine,
Process Maker and a bunch of Bitnami stacks in parallel without ever
noticing any problems caused by "poor performance". I am sure there is a
lot of make-believe specs planted by the provider, but having worked with
traditional servers I found VServers actually benchmarking quite well
considering the much lower price point in comparison. It sounds like my
case is actually pretty rare - but if it is not, do you think there is a
possibility for the hardware requirements coming down to the level of
similar products in the future?

As you can see I am testing a lot of cloud app deployment platforms, and I
am using the private server as tinker machine to test solutions I can then
recommend and implement for clients in a proper environment and security
precautions. A client won't invest in these, however, if cannot determine
and promote the usefulness of a tool that requires it though. And I can
only do that by fully customizing the system to the needs of the
stakeholder - which rarely works on demo servers unfortunately.

No worries, I am frequently running into this roadblock :) That said, kudos
to Sandstorm's customer service, I appreciate the prompt support! Hopefully
I will get the chance to work with your suite.

Best,
Daniel

**Comment by kentonv** [comment]
Hi @gorfreed,

The problem here isn't performance. The problem is that Linux-VServer is based on a modified version of Linux which does not support the Linux kernel's containerization features. These features are essential to Sandstorm because it uses containerization to isolate apps from each other, to give each app a consistent runtime environment, and to protect against security problems in apps. There's simply no way to run Sandstorm without containerization.

I do suggest that you contact your host and ask them which server options support Docker. Any server that supports Docker should be able to run Sandstorm. (Sandstorm does not use Docker, but it uses the same Linux kernel features, which are missing in Linux-VServer.)

**Comment by ocdtrekkie** [comment]
I think we should explicitly document that Sandstorm doesn't work on Linux-VServers.
